sparta/sparta/utils/SpartaSharedPointer.hpp

```diff
@@ -50,46 +71,28 @@ namespace sparta
     template <class PointerT>
     class SpartaSharedPointer
     {
+    public:
+        class SpartaWeakPointer;
+
+    private:
         /// Internal structure to keep track of the reference
         struct RefCount
         {
             explicit RefCount(PointerT * _p,
                               bool perform_delete = true) :
-                count(1),
                 p(_p),
-                perform_delete_(perform_delete){}
+                perform_delete(perform_delete)
+            {}
 
-            ~RefCount() {
-                if (perform_delete_) {
-                    delete p;
-                    p = nullptr;
-                }
-            }
+            ~RefCount() { p = nullptr; }
 
-            int32_t count;
-            PointerT * p = nullptr;
-            const bool perform_delete_;
+            int32_t count   {1};
+            int32_t wp_count{0}; // For weakpointers
+            PointerT          * p = nullptr;
+            void              * mem_block = nullptr;
```

Comment:
A good use case for `void *`!

Reply:
:laughing: The only time I've used it (I think) in my career!